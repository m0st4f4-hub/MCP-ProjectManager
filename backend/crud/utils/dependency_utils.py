# Task ID: Generated
# Agent Role: Agent (FixingCircularImports)
# Request ID: (Inherited from Overmind)
# Project: task-manager
# Timestamp: 2025-05-24T12:00:00Z

"""
Utility functions for working with task dependencies.
This file contains common functionality without circular imports.
"""

from typing import Any
from sqlalchemy.orm import Session
from sqlalchemy import and_
from ...models import TaskDependency
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select


def is_self_dependency(predecessor_project_id: str, predecessor_task_number: int,
                      successor_project_id: str, successor_task_number: int) -> bool:
    """
    Returns True if the predecessor and successor refer to the same task (self-dependency).
    """
    return predecessor_project_id == successor_project_id and predecessor_task_number == successor_task_number


async def get_direct_predecessors(db: AsyncSession, successor_project_id: str, successor_task_number: int) -> list[TaskDependency]:
    """
    Get all tasks that must complete before this task can start.
    This is a utility function used by validation logic.
    """
    result = await db.execute(select(TaskDependency).filter(
        and_(
            TaskDependency.successor_project_id == successor_project_id,
            TaskDependency.successor_task_number == successor_task_number
        )
    ))
    return result.scalars().all()
