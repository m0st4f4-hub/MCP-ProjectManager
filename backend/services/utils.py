"""
Transaction utility for service layer operations.
This module provides a context manager for handling database transactions.
"""

from contextlib import contextmanager, asynccontextmanager
from sqlalchemy.orm import Session
from sqlalchemy.exc import SQLAlchemyError
import logging

# Import service exceptions for consistent error handling
from backend.services.exceptions import ServiceError, EntityNotFoundError, DuplicateEntityError, ValidationError

logger = logging.getLogger(__name__)

@asynccontextmanager
async def service_transaction(db: Session, operation_name: str = "database operation"):
    """Context manager for database transactions in services."""
    try:
        # In an async context, beginning a transaction might be implicit or require specific handling
        # For now, assuming the session handles transaction boundaries correctly with await commit/rollback
        yield db
        await db.commit()
    except SQLAlchemyError as e:
        await db.rollback()
        # Log the error with operation_name for context
        # Example: logger.error(f"SQLAlchemyError during {operation_name}: {e}")
        # Convert SQLAlchemyError to a more generic ServiceError or a specific one if identifiable
        # This helps decouple service layer from direct SQLAlchemy exceptions if needed elsewhere
        if hasattr(e, 'orig') and e.orig is not None:
            if "UNIQUE constraint failed" in str(e.orig).lower():
                # Extract more specific details if possible, or use a generic message
                raise DuplicateEntityError(entity_name="Unknown", entity_id=operation_name) from e
            elif "NOT NULL constraint failed" in str(e.orig).lower():
                raise ValidationError(f"Required field missing during {operation_name}.") from e
        # Add more specific error mappings as needed
        # If e.orig doesn't exist or isn't a standard DBAPIError, or doesn't match known constraints
        raise ServiceError(f"Database error during {operation_name}: {str(e)}") from e
    except (ServiceError, EntityNotFoundError, DuplicateEntityError, ValidationError) as e:
        # Allow service-specific exceptions to propagate directly
        await db.rollback()
        raise
    except Exception as e:
        await db.rollback()
        # Log unexpected errors
        # Example: logger.error(f"Unexpected error during {operation_name}: {e}", exc_info=True)
        raise ServiceError(f"Unexpected error during {operation_name}: {str(e)}") from e
