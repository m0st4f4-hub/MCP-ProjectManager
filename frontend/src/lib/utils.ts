// import { Status } from '@/types';
import {
  getStatusAttributes,
  StatusID as CanonicalStatusID,
} from "@/lib/statusUtils";

export const formatDisplayName = (name?: string | null): string => {
  if (!name || name.trim() === "") {
    return "Unnamed";
  }

  // Replace underscores and hyphens with spaces, then capitalize each word
  const words = name.replace(/[_-]/g, " ").toLowerCase().split(" ");

  const capitalizedWords = words.map(
    (word) => word.charAt(0).toUpperCase() + word.slice(1),
  );

  return capitalizedWords.join(" ").trim();
};

// Example usage (can be removed or kept for testing)
/*
console.log(formatDisplayName("my_project_name")); // My Project Name
console.log(formatDisplayName("another-agent-name")); // Another Agent Name
console.log(formatDisplayName("SingleWord"));      // Singleword (Note: Could be improved to handle camelCase better if needed)
console.log(formatDisplayName("  leading space  ")); // Leading Space
console.log(formatDisplayName(null));                 // Unnamed
console.log(formatDisplayName(""));                   // Unnamed
*/

export const mapStatusToStatusID = (
  status: string | null | undefined,
): CanonicalStatusID => {
  if (!status || typeof status !== "string" || status.trim() === "") {
    return "To Do"; // Default for null, undefined, or empty/non-string if it somehow bypasses Status type
  }

  // Handle common display names case-insensitively
  const lowerStatus = status.toLowerCase();
  switch (lowerStatus) {
    case "to do":
    case "todo":
      return "To Do";
    case "in progress":
      return "In Progress";
    case "blocked":
      return "Blocked";
    case "completed":
      return "Completed";
    case "cancelled":
      return "Cancelled";
    // Add more explicit mappings if other common display names exist
    // e.g., case 'context acquired': return 'CONTEXT_ACQUIRED';
  }

  // Fallback: try to match the uppercased/underscored version directly as a CanonicalStatusID
  // This handles cases where task.status might already be a StatusID string (e.g., "EXECUTION_IN_PROGRESS")
  // or a custom status string that follows the StatusID format.
  const upperStatus = status.toUpperCase().replace(/\s+/g, "_");
  if (getStatusAttributes(upperStatus as CanonicalStatusID)) {
    // It's a known CanonicalStatusID.
    // Optionally, add a console.warn if 'status' wasn't identical to 'upperStatus'
    // to catch statuses that are almost StatusIDs but have different casing/spacing.
    if (status !== upperStatus && typeof console !== "undefined") {
      console.warn(
        `[mapStatusToStatusID] Mapped status "${status}" to CanonicalStatusID "${upperStatus}".`,
      );
    }
    return upperStatus as CanonicalStatusID;
  }

  // If no match after explicit cases and direct StatusID check, then it's unknown.
  if (typeof console !== "undefined") {
    console.warn(
      `[mapStatusToStatusID] Unknown status value: "${status}", defaulting to To Do.`,
    );
  }
  return "To Do";
};
