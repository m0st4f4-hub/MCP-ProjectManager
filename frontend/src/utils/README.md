# Utility Functions (`frontend/src/utils/`)

This directory is intended to store general-purpose utility functions that can be used across various parts of the frontend application.

These might include functions for data formatting, validation, calculations, or other helper routines that don't fit neatly into other categories like `lib` (which might be more for core/shared business logic or type-safe utilities) or specific component/module logic.

Currently, this directory is empty. As common, reusable utility functions are identified or developed, they can be placed here.

## Architecture Diagram
```mermaid
graph TD
    user((User)) -->|interacts with| frontend(Frontend)
    frontend -->|API requests| backend(Backend)
    backend -->|persists| database[(Database)]
    backend -->|integrates| mcp(MCP Server)
```

<!-- File List Start -->
## File List


<!-- File List End -->

