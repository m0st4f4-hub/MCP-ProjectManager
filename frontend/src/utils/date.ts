import { parseISO } from 'date-fns';

export function parseDate(value: string | number | Date): Date {
  if (value instanceof Date) {
    return value;
  }
  if (typeof value === 'number') {
    return new Date(value);
  }
  return parseISO(value);
}
