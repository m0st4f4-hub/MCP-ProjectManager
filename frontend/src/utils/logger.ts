const LEVELS = { info: 0, warn: 1, error: 2 } as const;
export type LogLevel = keyof typeof LEVELS;
const envLevel = process.env.NEXT_PUBLIC_LOG_LEVEL as LogLevel | undefined;
const currentLevel: LogLevel = envLevel && envLevel in LEVELS ? envLevel : 'info';
export function info(...args: unknown[]) {
  if (LEVELS[currentLevel] <= LEVELS.info) {
    console.info(...args);
  }
}
export function warn(...args: unknown[]) {
  if (LEVELS[currentLevel] <= LEVELS.warn) {
    console.warn(...args);
  }
}
export function error(...args: unknown[]) {
  if (LEVELS[currentLevel] <= LEVELS.error) {
    console.error(...args);
  }
}
export default { info, warn, error };
