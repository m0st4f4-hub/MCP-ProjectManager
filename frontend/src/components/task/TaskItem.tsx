// D:\mcp\task-manager\frontend\src\components\TaskItem.tsx
"use client";

import React, { useCallback, memo } from "react";
import {
  Box,
  Checkbox,
  HStack,
  VStack,
  useToast,
  Button,
} from "@chakra-ui/react";
import { defaultIconMap } from "../common/iconMap";
import { useProjectStore } from "@/store/projectStore";
import { useTaskStore } from "@/store/taskStore";
import { getDisplayableStatus, StatusID, StatusAttributeObject } from "@/lib/statusUtils";
import { TaskItemProps } from "./TaskItem.types";
import { getStatusAccentColor } from "@/components/task";
import { useTaskItemStyles } from "../useTaskItemStyles";
import TaskItemMainSection from "./TaskItemMainSection";
import TaskItemModals from "../TaskItemModals";
import { TaskStatus } from "@/types/task";

// Default mapping from status icon names to actual icon components.
const iconMap = defaultIconMap;

/**
 * @module TaskItem
 * @description
 * TaskItem is the primary component for rendering a single task. It orchestrates
 * the display of task details, actions, and manages UI states like expansion
 * or modal visibility for editing. It delegates rendering of specific sections
 * to sub-components like TaskItemMainSection and TaskItemDetailsSection.
 *
 * It displays a compact or full view of a task, including its completion status,
 * title, project, tags, and provides access to task actions.
 *
 * @example
 * // Example usage within a task list:
 * <TaskItem
 *   task={myTaskObject}
 *   compact={false}
 *   onClick={() => console.log('Task clicked')}
 * />
 *
 * @param {TaskItemProps} props - The props for the TaskItem component.
 * @param {ITask} props.task - The core task object containing all its data.
 * @param {boolean} [props.compact=false] - If true, renders a more compact version of the task item.
 * @param {React.CSSProperties} [props.style] - Custom styles to apply to the root element.
 * @param {() => void} [props.onClick] - Callback function invoked when the task item is clicked.
 */
const TaskItem: React.FC<TaskItemProps> = memo(
  function TaskItem({
    task, // The main task object.
    compact = false,
    style,
    onClick,
  }) {
    const projects = useProjectStore((state) => state.projects);
    const projectName = projects.find((p) => p.id === task.project_id)?.name; // Fetches project name from store

    const editTaskInStore = useTaskStore((state) => state.updateTask); // Zustand store action for updating tasks
    const archiveTask = useTaskStore((state) => state.archiveTask);
    const unarchiveTask = useTaskStore((state) => state.unarchiveTask);

    const toast = useToast(); // Chakra UI toast for notifications

    // Determine the current status ID, defaulting to "TO_DO" if not set.
    const currentStatusId = (task.status || "TO_DO") as StatusID;
    const statusInfo = getDisplayableStatus(currentStatusId, task.title); // Utility to get displayable status info (icon, color, etc.)

    const styles = useTaskItemStyles(currentStatusId, compact); // Custom hook to get styles based on status and compactness

    // Callback to toggle the task's completion status.
    // It optimistically updates the UI and then calls the store action.
    const handleToggleCompletion = useCallback(async () => {
      // Determine the new status based on the current status.
      const newStatus = task.status !== TaskStatus.COMPLETED ? TaskStatus.COMPLETED : TaskStatus.TO_DO;
      try {
        // Attempt to update the task in the store.
        await editTaskInStore(task.project_id, task.task_number, { status: newStatus });
      } catch {
        // If the update fails, show an error toast.
        toast({
          title: "Error updating status",
          status: "error",
          duration: 3000,
          isClosable: true,
        });
      }
    }, [task.project_id, task.task_number, task.status, editTaskInStore, toast]); // Dependencies for useCallback

    // Determine the accent color for the left border based on the task status.
    const currentAccentColor = getStatusAccentColor(
      currentStatusId,
      styles.accentToDo, // Style token for ToDo accent
      styles.accentInProgress,
      styles.accentBlocked,
      styles.accentPending,
      styles.accentCompleted,
      styles.accentDefault, // Default accent color
    );

    // Determine text color based on whether the task is archived.
    const textColor = task.is_archived
      ? styles.textDisabledColor // Disabled text color for archived tasks
      : styles.textPrimaryColor; // Primary text color for active tasks

    return (
      // Root Box element for the TaskItem
      <Box
        p={compact ? "spacing.2" : "spacing.4"} // Padding based on compact prop
        borderRadius="radii.md"
        borderWidth="borders.width.xs"
        borderStyle="solid"
        borderColor={styles.borderDecorativeColor} // Border color
        boxShadow="shadows.sm" // Default shadow
        transition="all 0.2s ease-in-out" // Smooth transition for hover effects
        position="relative" // For potential absolute positioned children
        overflow="hidden" // Ensures content respects border radius
        bg={
          // Background color based on completion status
          task.status === TaskStatus.COMPLETED
            ? styles.bgSurfaceElevatedColor // Elevated background for completed tasks
            : styles.bgSurfaceColor // Standard background for other tasks
        }
        borderLeftWidth="spacing.1" // Left border width for status accent
        borderLeftStyle="solid"
        borderLeftColor={currentAccentColor} // Status accent color
        _hover={{
          // Hover effects
          boxShadow: "shadows.md", // Increase shadow on hover
          transform: "translateY(-1px)", // Slight lift effect on hover
        }}
        style={style} // Apply custom inline styles if provided
        onClick={onClick} // Attach onClick handler if provided
      >
        {/* Horizontal stack for checkbox and main content */}
        <HStack alignItems="flex-start" width="100%" gap="spacing.3">
          {/* Checkbox for task completion */}
          <Box pt="spacing.0-5">
            <Checkbox
              isChecked={task.status === TaskStatus.COMPLETED} // Checked if task is completed
              onChange={handleToggleCompletion} // Handler to toggle completion
              size="lg" // Large size checkbox
              colorScheme={statusInfo?.colorScheme || "gray"} // Color scheme based on status
              aria-label={`Mark task ${task.title} as ${task.status === TaskStatus.COMPLETED ? "incomplete" : "complete"}`}
            />
          </Box>

          {/* Vertical stack for task details */}
          <VStack
            alignItems="flex-start" // Align items to the start (left)
            flexGrow={1} // Allow this stack to grow and fill available space
            minW={0} // Prevent overflow issues by setting a minimum width
            gap={compact ? "spacing.0-5" : "spacing.1"} // Gap between elements based on compact prop
          >
            {/* Main section of the task item, includes title, status, etc. */}
            <TaskItemMainSection
              task={task}
              projectName={projectName} // Pass down the fetched project name
              statusInfo={ // Pass down status information object
                (statusInfo !== undefined
                  ? statusInfo
                  : { // Fallback status info if getDisplayableStatus returns undefined
                      id: currentStatusId,
                      displayName: currentStatusId, // Default display name to status ID
                      colorScheme: 'gray', // Default color scheme
                      icon: undefined, // No icon by default
                      category: 'todo', // Default category
                      description: '', // No description
                      isTerminal: false, // Not terminal by default
                      isDynamic: false, // Not dynamic by default
                    }) as StatusAttributeObject // Use imported StatusAttributeObject
              }
              styles={styles} // Pass down styles from useTaskItemStyles
              textColor={textColor} // Pass down calculated text color
              iconMap={iconMap} // Pass down the populated iconMap
              currentStatusId={currentStatusId} // Pass down current status ID
              compact={compact} // Pass down compact prop
              editTaskInStore={editTaskInStore} // Pass down store action for editing
            />
          </VStack>
        </HStack>
        {/* Modals related to task actions (e.g., edit, details) */}
        {/* This component likely handles its own visibility based on internal state or props from TaskItem */}
        <TaskItemModals task={task} />
        <Button
          size="xs"
          onClick={(e) => {
            e.stopPropagation(); // Prevent task item click when clicking the button
            if (task.is_archived) {
              unarchiveTask(task.project_id, task.task_number);
            } else {
              archiveTask(task.project_id, task.task_number);
            }
          }}
        >
          {task.is_archived ? "Unarchive" : "Archive"}
        </Button>
      </Box>
    );
  }
);

// Set display name for better debugging in React DevTools
TaskItem.displayName = "TaskItem";

export default TaskItem;