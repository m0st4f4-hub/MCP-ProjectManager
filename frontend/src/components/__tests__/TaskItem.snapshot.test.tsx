import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import { render, TestWrapper } from '@/__tests__/utils/test-utils';
import { createMockTask } from '@/__tests__/factories';
import TaskItem from '../task/TaskItem';

vi.mock('@chakra-ui/react', async () => {
  const actual = await vi.importActual<typeof import('@chakra-ui/react')>('@chakra-ui/react');
  return {
    ...actual,
    useToast: () => vi.fn(),
    useColorModeValue: (light: any, dark: any) => light,
  };
});

describe('TaskItem snapshot', () => {
  it('renders consistently', () => {
    const task = createMockTask({
      project_id: 'project-id',
      task_number: 1,
      title: 'Snapshot Task',
      description: 'Snapshot description',
      created_at: '2024-01-01T00:00:00.000Z',
      updated_at: '2024-01-01T00:00:00.000Z',
      agent_id: null,
    });
    const { asFragment } = render(
      <TestWrapper>
        <TaskItem task={task} projectName="Project" onDeleteInitiate={() => {}} />
      </TestWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
