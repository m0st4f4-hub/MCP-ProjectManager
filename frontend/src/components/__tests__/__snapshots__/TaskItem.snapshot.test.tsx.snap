// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskItem snapshot > renders consistently 1`] = `
<DocumentFragment>
  <div
    class="css-1t1yp2j"
  >
    <div
      class="chakra-stack css-y10l5h"
    >
      <div
        class="css-1ydq2kx"
      >
        <label
          class="chakra-checkbox css-1577qb8"
        >
          <input
            aria-checked="false"
            aria-label="Mark task Snapshot Task as complete"
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-olidhu"
          />
        </label>
      </div>
      <div
        class="chakra-stack css-1w5ylqm"
      >
        <div
          class="chakra-stack css-1w5ylqm"
        >
          <div
            class="css-8atqhb"
          >
            <p
              class="chakra-text css-ru5nt8"
            >
              Snapshot Task
            </p>
          </div>
          <p
            class="chakra-text css-khfcme"
          >
            Snapshot description
          </p>
          <div
            class="chakra-stack css-wpcv6r"
          >
            <div
              class="chakra-stack css-h2mol3"
            >
              <span
                class="css-4owlnz"
              >
                <svg
                  class="chakra-icon chakra-icon css-l30hbq"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <g
                    fill="none"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-width="2"
                  >
                    <path
                      d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                    />
                    <path
                      d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                    />
                  </g>
                </svg>
                <span
                  class="css-1ny2kle"
                >
                  To Do
                </span>
              </span>
            </div>
            <div
              class="css-0"
            >
              <div
                class="chakra-spinner css-hno8u4"
              >
                <span
                  class="css-8b45rq"
                >
                  Loading...
                </span>
              </div>
              <p
                class="chakra-text css-6n7j50"
              >
                Loading task details...
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="chakra-button css-l7xuv7"
      type="button"
    >
      Archive
    </button>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
